File: 알고리즘__문제/프로그래머스/프로그래머스_lever1/연습문제/콜라즈추측.py
# def solution(num):
#     numbers = []
#     if num == 1:
#         print(numbers)
#         return len(numbers)

#     else:
#         if num % 2 == 0:
#             num = num // 2
#             numbers.append(num)
#             solution(num)

#         else:
#             num = num * 3 + 1
#             numbers.append(num)
#             solution(num)


def solution(num):
    count = 0  # cnt 맨 위에 정의하고
    while True:  # while문으로 반복돌리면 cnt가 0이 될 일이 없음...ㅠ
        if num == 1:
            return count
        count += 1  # 마지막 1도 추가해야함

        if num % 2:
            num *= 3
            num += 1
        else:
            num /= 2
        if count == 500:
            return -1


num = 6
print(solution(num))

File: 자바스크립트풀이/programmers/콜라즈추측.js
'use strict'

// function solution(num) {
//     while (num > 1) {

//         let cnt = 0;

//         if (cnt == 500){
//             return -1;
//         } else if (num == 1) {
//             return cnt;
//         } else if (num != 1) {
//             if (num % 2 == 0) {
//                 num / 2;
//                 cnt++;
//             } else {
//                 num * 3 + 1;
//                 cnt++;
//             }
//         }
//     }       
// }

function solution(num, count = 0) {
    return count === 500 
        ? -1 
        : num === 1
            ? count
            : solution(num % 2 ? num * 3 + 1 : num / 2, count + 1);
}


solution(8)

// ctrl + shift + i 로 개발자 도구 열기